src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -3152,40 +3152,109 @@ static Block.Builder lowerTypePattern(Block.Builder endNoMatchBlock, Block.Build
+                        if (isNarrowingPrimitiveConv(ps, pt) || isWideningPrimitiveConvWithCheck(ps, pt)
+                                || isWideningAndNarrowingPrimitiveConv(ps, pt)) {
+                            // e.g. int -> byte, narrowing
+                            // e,g. int -> float, widening with check
+                            // e.g. byte -> char, widening and narrowing
+                            MethodRef mref = convMethodRef(s, t);
+                            p = invoke(mref, target);
+                        }
+                        c = CoreOp.conv(targetType, target);
                     }
-                    String s = capitalize(st.toString());
-                    String t = capitalize(tt.toString());
-                    String mn = "is%sTo%sExact".formatted(s, t);
-                    JavaType exactConversionSupport = JavaType.type(ClassDesc.of("java.lang.runtime.ExactConversionsSupport"));
-                    MethodRef mref = MethodRef.method(exactConversionSupport, mn, BOOLEAN, st);
-                    p = currentBlock.op(invoke(mref, target));
-                } else {
-                    p = currentBlock.op(CoreOp.instanceOf(targetType, target));
+                } else if (s instanceof PrimitiveType ps) {
+                    // boxing conversions
+                    ClassType ct = (ClassType) t;
+                    ClassType box;
+                    if (ct.unbox().isEmpty()) {
+                        // e.g. int -> Number, boxing + widening
+                        box = ps.box().orElseThrow();
+                    } else {
+                        // e.g. byte -> Byte, boxing
+                        box = ct;
+                    }
+                    c = invoke(MethodRef.method(box, "valueOf", box, ps), target);
+                } else if (!Objects.equals(s, t)) {
+                    // reference to reference, but not identity
+                    // e.g. Number -> Double, narrowing
+                    // e.g. Short -> Object, widening
+                    p = CoreOp.instanceOf(targetType, target);
+                    c = CoreOp.cast(targetType, target);
                 }
 
-                currentBlock.op(conditionalBranch(p, nextBlock.successor(), endNoMatchBlock.successor()));
-
-                currentBlock = nextBlock;
+                if (p != null) {
+                    // p != null, we need to perform type check at runtime
+                    Block.Builder nextBlock = currentBlock.block();
+                    currentBlock.op(conditionalBranch(currentBlock.op(p), nextBlock.successor(), endNoMatchBlock.successor()));
+                    currentBlock = nextBlock;
+                }
 
-                if (patternWithPrimitive) {
-                    target = currentBlock.op(CoreOp.conv(targetType, target));
-                } else {
-                    target = currentBlock.op(CoreOp.cast(targetType, target));
+                if (c != null) {
```

Comment:
If `c` is `null` then i believe `p` is also `null`, but not the other way around. Can you place the `c != null` check before and embed in the `p != null` check to be clearer about the dependency?